File: LeetCode_Marathons/LCM6/Hard/Patching_Array/a.java
class Solution {
    public int minPatches(int[] nums, int n) {
        long dpSum = 0;
        int counter = 0;

        ArrayList<Long> arr = new ArrayList<>();
        for (int num: nums) arr.add((long) num);
        arr.add(((long) n) + 1);

        for (long num: arr) {
            while (dpSum + 1 < num) {
                if (dpSum >= (long) n) return counter;
                counter++;
                dpSum = dpSum + 1 + dpSum;
            }
            dpSum += num;
            if (dpSum >= (long) n) return counter;
        }

        return counter;
    }
}

File: LeetCode_Marathons/LCM6/Hard/Patching_Array/a.py
class Solution:
    def minPatches(self, nums: List[int], n: int) -> int:
        counter: int = 0
        dpSum: int = 0
        nums.append(n + 1)
        for num in nums:
            while dpSum + 1 < num:
                if dpSum >= n: return counter
                counter += 1
                dpSum = dpSum + 1 + dpSum
            dpSum += num
            if dpSum >= n: return counter
        return counter